honu.go

```diff
@@ -1,30 +1,185 @@
+	default:
+		return nil, fmt.Errorf("unhandled database scheme %q", dsn.Scheme)
+	}
+}
+
+// Close the database, allowing no further interactions.
+func (d *DB) Close() error {
+	return d.ldb.Close()
 }
 
-func Get(key []byte) (value []byte, err error) {
-	return nil, nil
+// Get the latest version of the object stored by the key.
+// TODO: provide read options to the underlying database.
+func (d *DB) Get(key []byte) (value []byte, err error) {
+	// Fetch the value from the database
+	if value, err = d.ldb.Get(key, nil); err != nil {
+		// TODO: should we wrap the leveldb error?
+		return nil, err
+	}
+
+	// Parse the object record to extract the value data
+	obj := new(pb.Object)
+	if err = proto.Unmarshal(value, obj); err != nil {
+		// TODO: better error message here
+		return nil, err
+	}
+
+	if obj.Tombstone() {
+		// The object is deleted, so return not found
+		return nil, leveldb.ErrNotFound
+	}
+
+	// Return the wrapped data
+	return obj.Data, nil
+
 }
 
-func Put(key, value []byte) (err error) {
+// Put a new value to the specified key and update the version.
+// TODO: provide write options to the underlying database.
+func (d *DB) Put(key, value []byte) (err error) {
+	// Get or Create the previous version
+	var data []byte
+	var obj *pb.Object
+	if data, err = d.ldb.Get(key, nil); err != nil {
+		if errors.Is(err, leveldb.ErrNotFound) {
+			obj = &pb.Object{
+				Key:       key,
+				Namespace: "default",
+			}
+		} else {
+			return err
+		}
+	} else {
+		obj = new(pb.Object)
+		if err = proto.Unmarshal(data, obj); err != nil {
+			return err
+		}
+	}
+
+	// Update the version with the new data
+	obj.Data = value
+	if err = d.vm.Update(obj); err != nil {
+		return err
+	}
+
+	// Put the version back onto disk
+	if data, err = proto.Marshal(obj); err != nil {
+		return err
+	}
+	if err = d.ldb.Put(key, data, nil); err != nil {
+		return err
+	}
+
 	return nil
 }
 
-func Delete(key []byte) (err error) {
+// Delete the object represented by the key, creating a tombstone object.
+// TODO: provide write options to the underlying database.
+func (d *DB) Delete(key []byte) (err error) {
+	var data []byte
+	if data, err = d.ldb.Get(key, nil); err != nil {
+		if errors.Is(err, leveldb.ErrNotFound) {
+			return nil
+		}
+		return err
+	}
+
+	// Unmarshal the version information
+	obj := new(pb.Object)
+	if err = proto.Unmarshal(data, obj); err != nil {
+		return err
+	}
+
+	// Don't save the data back to disk
+	obj.Data = nil
+
+	// Create a tombstone for the data
+	if err = d.vm.Delete(obj); err != nil {
+		return err
+	}
+
+	// Put the version back onto disk
+	if data, err = proto.Marshal(obj); err != nil {
+		return err
+	}
+	if err = d.ldb.Put(key, data, nil); err != nil {
+		return err
+	}
 	return nil
 }
 
-func Iter(prefix []byte) (i *iterator.Iterator, err error) {
-	return nil, nil
+// Iter over a subset of keys specified by the prefix.
+// TODO: provide better mechanisms for iteration.
```

Comment:
What kinds of improvements you have in mind? SQLite-related?

Reply:
I suppose I should have provided a better comment here, sorry about that - was copy and pasting very quickly!

All of the leveldb wrapper calls (especially this one) massively simplify the interactions with leveldb. E.g. all of them remove the ReadOptions or WriteOptions associated with the calls. And in this case, I was referring to the limitation of only allowing prefix iteration. The `ldb.NewIterator` call takes a `slice *util.Range` argument which is more flexible than simple prefixes (though we've primarily used prefixes in the past).

However, you're right that a lot of this simplification is required as we think through genericizing the lower-level embedded database. Iteration and even prefix iteration doesn't make a whole lot of sense for sqlite - since "iteration" in this context is just a SELECT query. I don't know if we can come up with a unified API or if we need a different API for Honu-sqlite vs Honu-leveldb. I am looking forward to exploring this issues as we flush this out as a more general open source project though!

Reply:
Gotcha, thank you for the explanation. I agree that supporting some of these standard ldb options  (at least the most popularly used ones) is a good choice; we'd like the transition from ldb to Honu to be as seamless for other users as possible, and we don't want them to feel unduly constrained.   Makes sense!